### PR TITLE
Navigation Grid Cleanup

### PR DESCRIPTION
**Pull request by Crauzer**

### CODE REVIEW

File: GameServerLib/Content/Navigation/NavigationGrid.cs [hunk 1]
```diff
@@ -0,0 +1,626 @@
...
+        {
+            if (a is IObjBuilding)
+            {
+                double rayDist = Math.Sqrt(CastRaySqr(b.GetPosition(), a.GetPosition()));
+                rayDist += a.CollisionRadius;
+                return (rayDist * rayDist) < (b.GetPosition() - a.GetPosition()).SqrLength();
+            }
+            else
+            {
+                return CastRaySqr(a.GetPosition(), b.GetPosition()) < (b.GetPosition() - a.GetPosition()).SqrLength();
```
Comment (danil179) [inline GameServerLib/Content/Navigation/NavigationGrid.cs]: Same problem.

**Review by danil179 — COMMENTED** [review]
It seems that you copy-pasted the last NavGrid so it still doesn't contain the changes. That's the only change that should be actually corrected with the correct code.

File: GameServerLib/Content/Navigation/NavigationGrid.cs (hunk 1 above)
Comment (Crauzer) [reply GameServerLib/Content/Navigation/NavigationGrid.cs]: What's the problem here ? Everything looks correct to me

Comment (danil179) [reply GameServerLib/Content/Navigation/NavigationGrid.cs]: > 
> 
> What's the problem here ? Everything looks correct to me

The problem is that you changed the current code: https://github.com/LeagueSandbox/GameServer/blob/8e9ab428855c22316af9978678d68f6c8d91b15e/GameServerLib/Content/NavGrid.cs#L553

File: GameServerLib/Content/Navigation/NavigationGrid.cs [hunk 2]
```diff
@@ -0,0 +1,626 @@
...
+                this.CellCountY = br.ReadUInt32();
+
+                this.Cells = new NavigationGridCell[this.CellCountX * this.CellCountY];
+                this.CellFlags = new ushort[this.CellCountX * this.CellCountY];
+
+                if (major == 2 || major == 3 || major == 5)
+                {
+                    for (int i = 0; i < this.Cells.Length; i++)
+                    {
+                        this.Cells[i] = new NavigationGridCell(br, i, out this.CellFlags[i]);
```
Comment (danil179) [inline GameServerLib/Content/Navigation/NavigationGrid.cs]: You declared 2 constructors with the same name for different versions. This is very confusing and causes code ambiguity.

Proposed solutions:
Sophisticated solution:
1. Create INavGridCellReader interface with `NavGridCell ReadNavGridCell(br, id)` function
2. Create abstract class that implements INavGridCellReader and call it NavGridCellReader
3. Create 2 class implementations (for each version) that implement the function ReadNavGridCell
4. Create a static function `INavGridCellReader GetReader(int major)` in NavGridCellReader that returns the correct reader for each version
5. Use it as `NavGridCellReader.GetReader(major).ReadNavGridCell(br, id)`

Simple hotfix solution (not recommended):
1. Add a switch to the constructor for different reading with each version

**Review by danil179 — CHANGES_REQUESTED** [review]
I have some problems with code ambiguity.

File: GameServerLib/Content/Navigation/NavigationGrid.cs (hunk 1 above)
Comment (Crauzer) [reply GameServerLib/Content/Navigation/NavigationGrid.cs]: Fixed 👌

File: GameServerLib/Content/Navigation/NavigationGrid.cs (hunk 2 above)
Comment (Crauzer) [reply GameServerLib/Content/Navigation/NavigationGrid.cs]: I implemented a solution to this in a commit that's better than creating redundant classes or interfaces

**Comment by danil179** [comment]
Opened issue #943 about the navigation grid flags storage. (They should be stored in NavGridCell and not in the NavGrid itself).

File: GameServerLib/Content/Navigation/NavigationGrid.cs (hunk 2 above)
Comment (danil179) [reply GameServerLib/Content/Navigation/NavigationGrid.cs]: The solution isn't really a great one (empty constructor, it should be private at least). I accept the changes.

File: GameServerLib/Content/Navigation/NavigationGridCell.cs [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+        public short Y { private set; get; }
+        public float AdditionalCost { private set; get; }
+        public float HintAsGoodCell { private set; get; }
+        public uint AdditionalCostRefCount { private set; get; }
+        public uint GoodCellSessionId { private set; get; }
+        public float RefHintWeight { private set; get; }
+        public short ArrivalDirection { private set; get; }
+        public short[] RefHintNode { private set; get; } = new short[2];
+
+        public NavigationGridCell() { }
```
Comment (danil179) [inline GameServerLib/Content/Navigation/NavigationGridCell.cs]: Make this private.

**Review by danil179 — CHANGES_REQUESTED** [review]
Make the constructor private to avoid future problems if possible.

File: GameServerLib/Content/Navigation/NavigationGridCell.cs (hunk 3 above)
Comment (Crauzer) [reply GameServerLib/Content/Navigation/NavigationGridCell.cs]: Done

**Review by danil179 — APPROVED** [review]
Approved. Please open an issue describing the problem you have solved.

**Pull request merged**